python/lsst/daf/persistence/butler.py

```diff
@@ -229,35 +437,49 @@ def getKeys(self, datasetType=None, level=None):
         @returns (dict) valid data id keys; values are corresponding types.
         """
         datasetType = self._resolveDatasetTypeAlias(datasetType)
-        return self.repository.getKeys(datasetType, level)
+
+        keys = None
+        for repoData in self._repos.inputs():
+            if not tag or len(tag.intersection(repoData.tags)) > 0:
+                keys = repoData.repo.getKeys(datasetType, level)
+                # An empty dict is a valid "found" condition for keys. The only value for keys that should
+                # cause the search to continue is None
+                if keys is not None:
+                    break
+        return keys
 
 
     def queryMetadata(self, datasetType, format=None, dataId={}, **rest):
         """Returns the valid values for one or more keys when given a partial
         input collection data id.
 
-        @param datasetType (str)    the type of dataset to inquire about.
-        @param key (str)            a key giving the level of granularity of the inquiry.
-        @param format (str, tuple)  an optional key or tuple of keys to be returned.
-        @param dataId (dict)        the partial data id.
-        @param **rest               keyword arguments for the partial data id.
+        @param datasetType (str) the type of dataset to inquire about.
+        @param key (str) a key giving the level of granularity of the inquiry.
+        @param format (str, tuple) an optional key or tuple of keys to be returned.
+        @param dataId (DataId, dict) the partial data id.
+        @param **rest keyword arguments for the partial data id.
         @returns (list) a list of valid values or tuples of valid values as
         specified by the format (defaulting to the same as the key) at the
         key's level of granularity.
         """
 
         datasetType = self._resolveDatasetTypeAlias(datasetType)
-        dataId = copy.copy(dataId)
+        dataId = DataId(dataId)
         dataId.update(**rest)
```

Comment:
I'm not sure it would be correct for the tag to be updated by values in ``**rest**``. DataId allows for the ``tag`` key as well as tag var (which could be confusing, but to disallow ``tag`` as a key also seems potentially confusing to me). Do you have a use case for this?